### PR TITLE
Add network connectivity detection service

### LINKED ISSUE #272: Network connectivity detection service
### Description
Implement a service to monitor the device's network connectivity state (online/offline) in real-time. This is foundational for the offline map feature, as it determines when to use cached tiles versus fetching fresh data from Mapbox servers.

The service should provide a reactive stream of connectivity states that can be observed by the map screen and other components. This enables the app to seamlessly transition between online and offline modes without user intervention.

### Why This is Needed
The parent issue (#271) requires the map to detect offline scenarios and switch to cached tiles automatically. Without connectivity detection, the app cannot determine when to use cached data versus when to fetch from the network.

### Acceptance Criteria
- A connectivity service monitors network state using Android's `ConnectivityManager`
- The service exposes the current connectivity state as a Kotlin Flow or LiveData for reactive observation
- The service correctly detects transitions between online and offline states
- The service handles edge cases (airplane mode, WiFi without internet, mobile data toggle)
- Unit tests verify correct state emission for various network scenarios
- Tests achieve 80%+ code coverage

### Testing Requirements
- Unit tests for connectivity state changes
- Mock `ConnectivityManager` to simulate different network conditions
- Verify Flow emissions for online/offline transitions
- Test edge cases (no connectivity callback support on older APIs)

### Dependencies
None (foundational PR)

### PR DESCRIPTION
**Pull request by ziyad-elarroussi**
## Description
Implement network connectivity detection service to monitor device network state in real-time. This service provides a reactive Flow-based API that other components can observe to determine when the device is online or offline.

This is the first step toward implementing offline map functionality, providing the foundation for subsequent PRs that will handle tile caching and offline map rendering.

**Related Issue:** closes #272.

---

## Changes

**Implementation (182 lines):**
  - Created ConnectivityService interface with Flow-based reactive API for observing connectivity state
  - Implemented ConnectivityServiceImpl using Android's ConnectivityManager with NetworkCallback for real-time updates
  - Added ConnectivityState data class to represent connection status and network type
  - Added NetworkType enum to distinguish between WiFi, Cellular, Ethernet, and Other connections
  - Implemented proper internet validation (checks NET_CAPABILITY_VALIDATED, not just network availability)
  - Created ConnectivityServiceProvider singleton following project's provider pattern
  - Handles edge cases: airplane mode, WiFi without internet, mobile data toggle

**Tests:**
  - Added comprehensive unit tests with Robolectric for Android framework mocking
  - 22 test cases covering all connectivity scenarios (online/offline, network type detection, edge cases)
  - Tests for singleton provider behavior (getInstance, setInstance, clearInstance)
  - All tests passing with 86% instruction coverage and 91% branch coverage (JaCoCo)

---

## Testing
- [X] Unit tests pass
- [X] Instrumentation tests pass - N/A - no instrumentation tests needed for this component
- [X] Manual testing completed
- [X] Coverage maintained (≥80%)

**Test summary:** Added 22 unit tests covering connectivity state detection, network type identification, Flow emissions, singleton provider behavior, and edge cases like airplane mode and WiFi without  internet validation.

---

## Screenshots/Videos
 N/A - This is a service layer component with no UI changes.
 
---

## Checklist
- [x] Sufficient documentation and minimal inline comments
- [x] All tests passing
- [x] No new warnings
- [x] If related issue exists: issue has all labels, fields, and milestone filled

### CODE REVIEW

**Review by e-bernier — CHANGES_REQUESTED** [review]
Great work on the connectivity service implementation! The reactive Flow design is impressive. One minor note: adding error handling  in the callbackFlow for getCurrentConnectivityState would make it even more robust. Otherwise, solid contribution!

File: app/src/main/java/com/swent/mapin/model/network/ConnectivityService.kt [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+ * @param context Application context
+ */
+class ConnectivityServiceImpl(context: Context) : ConnectivityService {
+  private val connectivityManager =
+      context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
+
+  override val connectivityState: Flow<ConnectivityState> =
+      callbackFlow {
+            // Send initial state
+            trySend(getCurrentConnectivityState())
```
[the remark this answers is not in the document]
Comment (ziyad-elarroussi) [reply app/src/main/java/com/swent/mapin/model/network/ConnectivityService.kt]: Changed: I've added try-catch blocks around all   getCurrentConnectivityState() calls in the callbackFlow. I used   try-catch instead of .onFailure{} because trySend() returns ChannelResult  (not Result), so .onFailure{} isn't available on it.

File: app/src/test/java/com/swent/mapin/model/network/ConnectivityServiceTest.kt [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    `when`(mockNetworkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET))
+        .thenReturn(true)
+    `when`(mockNetworkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED))
+        .thenReturn(true)
+    `when`(mockNetworkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI))
+        .thenReturn(false)
+    `when`(mockNetworkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR))
+        .thenReturn(false)
+    `when`(mockNetworkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET))
+        .thenReturn(false)
```
[the remark this answers is not in the document]
Comment (ziyad-elarroussi) [reply app/src/test/java/com/swent/mapin/model/network/ConnectivityServiceTest.kt]: Changed:   refactored this by adding a setupConnectedNetwork() helper function with sensible defaults for all capability mocks. Tests now call this helper with only the parameters they need to override.

**Review by e-bernier — APPROVED** [review]
This is now all good for me, good job.
